### PR TITLE
Feature/pumba support

### PR DESCRIPTION
**Pull request by slnowak**
Ok, this would be probably first approach to tackle this, since I don't really have much time right now to improve it.
It's not yet fully done, but please at least review it.

### CODE REVIEW

**Comment by slnowak** [comment]
Not really sure why it failed on CI, i will try to fix it during the weekend

**Comment by bsideup** [comment]
Hey @slnowak,

Any luck?

**Comment by slnowak** [comment]
@bsideup I've abandoned it because of lack and response. If you're still interested, I could resurrect the idea.

**Comment by bsideup** [comment]
@slnowak you did a great job and for sure we wish to see the final result! :)

Do you need some help from our side?

**Comment by slnowak** [comment]
I will try to find some time to work on it this week and give you some updates.

**Comment by bsideup** [comment]
Great, thanks :)

**Comment by slnowak** [comment]
@bsideup 
Did some refactoring (lost almost all package private classes :(, but at least package structure is more readable now). 
Tests should now pass on travis. However, I think it should be in a separate repo.

**Comment by bsideup** [comment]
Hi @slnowak,

Looks impressive! I'll do my best to find time to re-review it soon 👍 
I especially like the DSL part :)

If you want, I can create a repo under "testcontainers" org for you, so that you can put this module there and even take the ownership :)

We already did that for a couple of modules:
https://github.com/testcontainers/testcontainers-java-module-mariadb
https://github.com/testcontainers/testcontainers-java-module-dynalite
https://github.com/testcontainers/testcontainers-java-module-vault/pull/1

Or we can do the merge and extract it ourselves. Just let me know what works best for you!

Thanks!

**Comment by rnorth** [comment]
Awesome, thank you @slnowak! I'm looking forward to being able to add this
to the Testcontainers family 😀

>

**Comment by rnorth** [comment]
Reviewing and hoping to have a good play with this on a flight today. Looking pretty great so far though :)

**Comment by rnorth** [comment]
This is pretty amazing to use - thanks @slnowak! Aside from the trivial points mentioned above, the only other thing it might benefit from is Javadocs and some examples, perhaps in docs. Hopefully this would help people discover the DSL. This can wait, though.

So that we can land this, I've created a repo, [testcontainers/testcontainers-java-module-pumba](https://github.com/testcontainers/testcontainers-java-module-pumba), and made you an admin of it. If you have capacity to move this to there that'd be great - if not, please let us know and we'll try and help out. The testcontainers dynalite/mariadb repos are probably both reasonable examples to base off, for repo structure etc.

How does that sound?

Thanks again!

**Comment by slnowak** [comment]
I will provide fixes and move entire codebase to newly created repo during the weekend ;)

File: modules/chaostesting/src/main/java/org/testcontainers/PumbaContainer.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    private void executeCommand(PumbaCommand command) {
+        final String evaluatedCommand = command.evaluate();
+        setCommand(evaluatedCommand);
+        log.info("Executing pumba container with command \"{}\"", evaluatedCommand);
+        super.start();
+    }
+
+    private static class DoNotCheckStartup extends StartupCheckStrategy {
```
Comment (slnowak) [inline modules/chaostesting/src/main/java/org/testcontainers/PumbaContainer.java]: I've already tried it and it didn't work well with pubma. Maybe I will take another look on that.

**Comment by rnorth** [comment]
We're shortly going to be merging #574, which changes our build system to Gradle. This is in part intended to make contributions of modules easier (per #564), but unfortunately means that for a short while your PR is going to show merge conflicts with the master branch.

I just want to let you know we don't want to create new work for you: *we'll take care of the merge conflicts shortly*. Please don't worry - we're grateful for your PR and want to help integrate it soon. Thank you.

**Comment by slnowak** [comment]
No worries, I can manage merge conflict myself if that's necesary. I steel need to fix a few things, but unfortunately I can only contribute during the weekends (well, not this one unfortunately). I will try to 'close' the PR next weekend/

**Comment by bearrito** [comment]
Any update on this effort? I mentioned this in the other repo, but I've a developer resource I can allocate to this if I understand the best way to contribute and outstanding issues.

**Comment by slnowak** [comment]
@bearrito you can check the PR, see what's missing in the API. Any suggestions regarding improving the api are welcome. Please check integration tests to see the usage.

**Comment by bearrito** [comment]
@rnorth Is this still contingent on the repo's switch to gradle?
If I addressed the failing status checks above does that help get this in for a release?

**Comment by rnorth** [comment]
Sorry, @bearrito, for the slow reply. We've switched over to Gradle now, as has @slnowak for this PR. 

I think now's the time for us to review and hopefully merge very soon.

If you have someone who could try it out before merging, that would be fantastic! Real-world usage would provide useful context and ideas for improvements. I suspect it would be helpful to have all comments directly on this PR, and we can together decide which items need to be looked at pre-merge and which should be split off into separate tickets. How does that sound, @bearrito, @slnowak?

Other than building from source, it should also be possible to use a jitpack build of the pumba branch, as described here: https://github.com/testcontainers/testcontainers-java/blob/master/docs/usage.md#jitpack-unreleased-versions

Please let us know your thoughts!

**Comment by bearrito** [comment]
We will start on this next week.

**Comment by bsideup** [comment]
@bearrito @slnowak Any news? :)

**Comment by slnowak** [comment]
Waiting for either review or feedback from @bearrito

**Comment by bsideup** [comment]
Ok, this PR was "in review" for too long, let's do our best to deliver it 👍 

@slnowak @bearrito I'll go thru it one more time, is it ready for the final review?

Before we check the code, I would like to raise a few questions to understand the scope:
1) if we use Network-based commands, does it affect only selected containers or all of them?
2) Is it even possible to select container targets for Pumba?
3) Can be explicitly protect the targets with label? Wee assign `DockerClientFactory.DEFAULT_LABELS` to **every** container (except our internal ones which **must not** be affected) started by Testcontainers

These are important because if you run your tests on CI environment other projects (and Jenkins-inside-container) might not be happy :D

File: modules/chaostesting/src/main/java/org/testcontainers/client/commandparts/SupportedTimeUnit.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
+package org.testcontainers.client.commandparts;
+
+/**
+ * Created by novy on 01.01.17.
+ */
+public enum SupportedTimeUnit {
```
Comment (bsideup) [inline modules/chaostesting/src/main/java/org/testcontainers/client/commandparts/SupportedTimeUnit.java]: Do we really have to be that strict? WDYT about using Java's `Duration` and throwing error at runtime instead?

File: modules/chaostesting/src/main/java/org/testcontainers/executables/PumbaContainer.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        addEnv("DOCKER_HOST", String.format("unix://%s", DOCKER_SOCKET_CONTAINER_PATH));
+    }
+
+    @SneakyThrows
+    private void fetchIPRouteImage() {
+        new RemoteDockerImage(IP_ROUTE_DOCKER_IMAGE).get();
+    }
+
+    private void setupLogging() {
+        withLogConsumer(frame -> log.debug("Pumba container: \"{}\"", frame.getUtf8String()));
```
Comment (bsideup) [inline modules/chaostesting/src/main/java/org/testcontainers/executables/PumbaContainer.java]: plz use `new Slf4JConsumer(log)`

File: modules/chaostesting/src/main/java/org/testcontainers/executables/PumbaContainer.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import static org.testcontainers.containers.BindMode.READ_WRITE;
+
+/**
+ * Created by novy on 31.12.16.
+ */
+@Slf4j
+class PumbaContainer extends GenericContainer<PumbaContainer> {
+
+    private static final String PUMBA_DOCKER_IMAGE = "gaiaadm/pumba:0.4.7";
+    private static final String IP_ROUTE_DOCKER_IMAGE = "gaiadocker/iproute2:3.3";
```
Comment (bsideup) [inline modules/chaostesting/src/main/java/org/testcontainers/executables/PumbaContainer.java]: I think we already have this logic, check `DockerClientConfigUtils`

File: modules/chaostesting/src/main/java/org/testcontainers/executables/PumbaContainer.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                        .run("echo 'set -e' >> /docker_entrypoint.sh")
+                        .run("echo 'exec gosu root:root \"$@\"' >> /docker_entrypoint.sh")
+                );
+    }
+
+    private static String dockerSocketHostPath() {
+        return SystemUtils.IS_OS_WINDOWS ? "/" + DOCKER_SOCKET_HOST_PATH : DOCKER_SOCKET_HOST_PATH;
+    }
+
+    private static class FailOnlyOnErrorExitCode extends StartupCheckStrategy {
```
Comment (bsideup) [inline modules/chaostesting/src/main/java/org/testcontainers/executables/PumbaContainer.java]: AFAIK `OneShotStartupCheckStrategy` does the same

**Comment by bsideup** [comment]
I have a feeling that the PR is a bit over-engineered but I like it anyways, would love to see some things simplified tho. Especially some one method classes :)

**Comment by slnowak** [comment]
@bsideup I will adress the pr over the weekend + I will check how it behaves on multiple networks + stuff

File: modules/chaostesting/src/main/java/org/testcontainers/client/commandparts/SupportedTimeUnit.java (hunk 2 above)
Comment (slnowak) [reply modules/chaostesting/src/main/java/org/testcontainers/client/commandparts/SupportedTimeUnit.java]: Yup, will get rid of it in favour of plain old Duration

File: modules/chaostesting/src/main/java/org/testcontainers/executables/PumbaContainer.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    @SneakyThrows
+    private void fetchIPRouteImage() {
+        new RemoteDockerImage(IP_ROUTE_DOCKER_IMAGE).get();
+    }
+
+    private void setupLogging() {
+        withLogConsumer(frame -> log.debug("Pumba container: \"{}\"", frame.getUtf8String()));
+    }
+
+    private static ImageFromDockerfile buildPumbaDockerImage() {
```
[the remark this answers is not in the document]
Comment (slnowak) [reply modules/chaostesting/src/main/java/org/testcontainers/executables/PumbaContainer.java]: I will check it - afair it was a workaround to be able to run it on CI

File: modules/chaostesting/src/main/java/org/testcontainers/executables/PumbaContainer.java (hunk 5 above)
Comment (slnowak) [reply modules/chaostesting/src/main/java/org/testcontainers/executables/PumbaContainer.java]: If you run recurring command, pumba will be still running for a while, OneShotStartupCheckStrategy does not fit here

**Comment by slnowak** [comment]
@bsideup 
Regarding pumba targets:
By default it interacts with all the containers running on docker, which is probably a no-go. I think we should restrict it only to containers spawned by testcontainers.
Unfortunately, there's not much we can do - see https://github.com/alexei-led/pumba/issues/41 - you either need to play around with container names or you can annotate all the containers that should be ignored by pumba with `--label com.gaiaadm.pumba.skip=true` (which is of course not possible and not desirable). 
Maybe we could submit a PR to pumba so that it takes requested labels into account?

**Comment by bsideup** [comment]
Hi @slnowak,

Yes, a PR to Pumba sounds like the way to go 👍 
Currently, they support "excludes" with `com.gaiaadm.pumba.skip`, and a PR might introduce "includes".

**Comment by slnowak** [comment]
Ok, I will work on pumba PR in a day or two.

**Comment by bsideup** [comment]
@slnowak thanks a lot for your efforts on bringing this! Amazing commitment 💪

**Comment by slnowak** [comment]
For reference https://github.com/alexei-led/pumba/issues/86
I've started some development to integrate this but let's wait for the pumba author's response.

**Comment by rnorth** [comment]
Let’s keep this one open!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by rnorth** [comment]
I'm afraid it's time to close this PR - sorry that we didn't manage to land it in the end. 

We've just released a [Toxiproxy module](https://www.testcontainers.org/modules/toxiproxy/) for resilience testing, which is arguably a less sophisticated model, but is much simpler from an implementation perspective.

I hope the Toxiproxy module meets the needs of anyone following this PR.

@slnowak thanks for your efforts, and sorry if you feel the work was in vain. Your idea, to be able to conduct chaos testing with Testcontainers, is still going to make a big difference for Testcontainers users and we're grateful for that.

**Pull request closed without merging**